File: src/runtime/boxes/any/bb_any.cs
// bb_any.cs — ANY: match one char if in charset
// Mirrors src/runtime/boxes/any/bb_any.c
//
// α: if cursor >= Length or char not in charset → ω;  cursor++; γ(1 char)
// β: cursor--; ω

namespace Snobol4.Runtime.Boxes;

public sealed class bb_any : IByrdBox
{
    private readonly string _chars;
    public bb_any(string chars) { _chars = chars ?? ""; }

    public Spec Alpha(MatchState ms)
    {
        if (ms.Cursor >= ms.Length)           return Spec.Fail;
        if (!ms.CharInSet(ms.Cursor, _chars)) return Spec.Fail;
        var result = Spec.Of(ms.Cursor, 1);
        ms.Cursor++;
        return result;
    }

    public Spec Beta(MatchState ms)
    {
        ms.Cursor--;
        return Spec.Fail;
    }
}

File: src/runtime/boxes/any/bb_any.java
package bb;

/**
 * bb_any.java — ANY: match one char if in set
 * Port of bb_any.c / bb_any.s
 *
 *   ANY_α:  if (Δ>=Ω || chars.indexOf(Σ[Δ])<0)  goto ANY_ω;
 *           ANY=spec(Σ+Δ,1); Δ++;                goto ANY_γ;
 *   ANY_β:  Δ--;                                 goto ANY_ω;
 */
public class bb_any extends bb_box {
    private final String chars;

    public bb_any(MatchState ms, String chars) { super(ms); this.chars=chars; }

    @Override public Spec α() {
        if (ms.delta >= ms.omega) return null;
        if (chars.indexOf(ms.sigma.charAt(ms.delta)) < 0) return null;
        return new Spec(ms.delta++, 1);
    }

    @Override public Spec β() { ms.delta--; return null; }
}
